Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeInfoWrapper.cs

```diff
@@ -0,0 +1,562 @@
+    ///     so there will be little bugs.  This bug is also resolved in the provided wrappers.
+    ///
+    /// This class can also be cast to ComTypes.ITypeInfo for raw access to the underlying type information
+    /// </remarks>
+    public sealed class TypeInfoWrapper : ITypeInfoInternalSelfMarshalForwarder, IDisposable
+    {
+        private DisposableList<TypeInfoWrapper> _cachedReferencedTypeInfos;
+        private IntPtr _target_ITypeInfoPtr;
+        private ITypeInfoInternal _target_ITypeInfo;
+        private ITypeInfoInternal _target_ITypeInfoAlternate;
+        private bool _target_ITypeInfo_IsRefCounted;
+
+        public ITypeLibInternalSelfMarshalForwarder Container { get; private set; }
+        public int ContainerIndex { get; private set; }
+        public bool HasModuleScopeCompilationErrors { get; private set; }
+        public bool HasVBEExtensions { get; private set; }
+        public ComTypes.TYPEATTR CachedAttributes { get; private set; }
+        public bool HasSimulatedContainer { get; private set; }
+        public bool IsUserFormBaseClass { get; private set; }
+
+        public TypeInfoFunctionCollection Funcs;
+        public TypeInfoVariablesCollection Vars;
+        public TypeInfoImplementedInterfacesCollection ImplementedInterfaces;
+
+        // some helpers
+        public string Name => CachedTextFields._name;
+        public string DocString => CachedTextFields._docString;
+        public int HelpContext => CachedTextFields._helpContext;
+        public string HelpFile => CachedTextFields._helpFile;
+        public string ProgID => ContainerName + "." + CachedTextFields._name;
+        public Guid GUID => CachedAttributes.guid;
+        public TYPEKIND_VBE TypeKind => (TYPEKIND_VBE)CachedAttributes.typekind;
+        public bool HasPredeclaredId => CachedAttributes.wTypeFlags.HasFlag(ComTypes.TYPEFLAGS.TYPEFLAG_FPREDECLID);
+        public ComTypes.TYPEFLAGS Flags => CachedAttributes.wTypeFlags;
+        public string ContainerName => Marshal.GetTypeLibName(Container);
+
+        // Constants inside VBA components are exposed via the ITypeInfo, but there names are not reported correctly.
+        // Their names all appear with a DispID of MEMBERID_NIL.  In order to try to make VBA type infos more agreeable to the specifications,
+        // we make up some unique names for these constants, and create unique DispIDs for them at runtime.  
+        // This will help some ITypeInfo consumers that may not like the unnamed fields. 
+        // Currently this is achieved by defining a range in the 32-bit space used by DispIDs that is unlikely to conflict with 
+        // any normal DispIDs assigned by VBA, and one would think unlikely to be used by custom VB_UserMemId attributes. 
+        // The range chosen allows for 65536 constants, starting at _ourConstantsDispatchMemberIDRangeStart.
+        // generated names are in the format "_constantFieldId" + Index (where index is the index into GetVarDesc)
```

Comment:
The range won't conflict with those weird negative DIspIDs used by VB6, right? (ref [chat](https://chat.stackexchange.com/transcript/14929?m=48078094#48078094))

Reply:
should be well out of that range.